### DISCUSSION TITLE
Errors in IPAM deployment

### DISCUSSION THREAD

**Discussion by mgasaway67** [Azure/ipam#64] [Q&A]
I am receiving errors in the IPAM deployment pretty early on. I am deploying from vscode running on mac.

Error:
New-MgOauth2PermissionGrant_CreateExpanded: /Users/{username}/Scripts/ipam-main/deploy/deploy.ps1:492:7
Line |
 492 |        New-MgOauth2PermissionGrant `
     |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Code: generalException Message: Location header not present in redirection response. 
INFO: Admin consent for Microsoft Graph API permissions granted successfully
INFO: Granting admin consent to the IPAM UI application for exposed API from the IPAM Engine application
New-MgOauth2PermissionGrant_CreateExpanded: /Users/{username}/Scripts/ipam-main/deploy/deploy.ps1:508:5
Line |
 508 |      New-MgOauth2PermissionGrant `
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Code: generalException Message: Location header not present in redirection response. 
INFO: Admin consent for IPAM Engine exposed API granted successfully
INFO: Granting admin consent for Azure Service Management API permissions assigned to IPAM Engine application
New-MgOauth2PermissionGrant_CreateExpanded: /Users/{username}/Scripts/ipam-main/deploy/deploy.ps1:526:5
Line |
 526 |      New-MgOauth2PermissionGrant `
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Code: generalException Message: Location header not present in redirection response. 
INFO: Admin consent for Azure Service Management API permissions granted successfully
INFO: Deploying IPAM bicep templates
ERROR: Unable to deploy Azure IPAM solution due to an exception, see ../logs/deploy_20230104104142AM.log and ../logs/error_20230104104142AM.log for detailed information!

Log File:
**********************
PowerShell transcript start
Start time: 20230104104142
Username: {username}
RunAs User: {username}
Configuration Name: 
Machine: --- (Unix 10.15.7)
Host Application: /usr/local/microsoft/powershell/7/pwsh.dll
Process ID: 47401
PSVersion: 7.3.1
PSEdition: Core
GitCommitId: 7.3.1
OS: Darwin 19.6.0 Darwin Kernel Version 19.6.0: Tue Jun 21 21:18:39 PDT 2022; root:xnu-6153.141.66~1/RELEASE_X86_64
Platform: Unix
PSCompatibleVersions: 1.0, 2.0, 3.0, 4.0, 5.0, 5.1.10032.0, 6.0.0, 6.1.0, 6.2.0, 7.0.0, 7.1.0, 7.2.0, 7.3.1
PSRemotingProtocolVersion: 2.3
SerializationVersion: 1.1.0.1
WSManStackVersion: 3.0
**********************

NOTE: IPAM Deployment Type: Full
INFO: Fetching Tenant ID from Azure PowerShell SDK
INFO: Fetching Azure Cloud type from Azure PowerShell SDK
INFO: Validating Azure Region selected for deployment
INFO: Azure Region validated successfully
INFO: Creating Azure IPAM UI Application
INFO: Creating Azure IPAM Engine Application
INFO: Updating Azure IPAM Engine API Endpoint
INFO: Updating Azure IPAM UI Application Resource Access
INFO: Creating Azure IPAM UI Service Principal
INFO: Creating Azure IPAM Engine Service Principal
INFO: Creating Azure IPAM Engine Secret
INFO: Azure IPAM Engine & UI Applications/Service Principals created successfully
INFO: Logging in to Microsoft Graph
INFO: Granting admin consent for Microsoft Graph API permissions assigned to IPAM UI application

New-MgOauth2PermissionGrant_CreateExpanded: /Users/{username}/Scripts/ipam-main/deploy/deploy.ps1:492:7
Line |
 492 |        New-MgOauth2PermissionGrant `
     |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Code: generalException Message: Location header not present in redirection response.

New-MgOauth2PermissionGrant_CreateExpanded: /Users/{username}/Scripts/ipam-main/deploy/deploy.ps1:492:7
Line |
 492 |        New-MgOauth2PermissionGrant `
     |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Code: generalException Message: Location header not present in redirection response. 

INFO: Admin consent for Microsoft Graph API permissions granted successfully
INFO: Granting admin consent to the IPAM UI application for exposed API from the IPAM Engine application

New-MgOauth2PermissionGrant_CreateExpanded: /Users/{username}/Scripts/ipam-main/deploy/deploy.ps1:508:5
Line |
 508 |      New-MgOauth2PermissionGrant `
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Code: generalException Message: Location header not present in redirection response.

New-MgOauth2PermissionGrant_CreateExpanded: /Users/{username}/Scripts/ipam-main/deploy/deploy.ps1:508:5
Line |
 508 |      New-MgOauth2PermissionGrant `
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Code: generalException Message: Location header not present in redirection response. 

INFO: Admin consent for IPAM Engine exposed API granted successfully
INFO: Granting admin consent for Azure Service Management API permissions assigned to IPAM Engine application

New-MgOauth2PermissionGrant_CreateExpanded: /Users/{username}/Scripts/ipam-main/deploy/deploy.ps1:526:5
Line |
 526 |      New-MgOauth2PermissionGrant `
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Code: generalException Message: Location header not present in redirection response.

New-MgOauth2PermissionGrant_CreateExpanded: /Users/{username}/Scripts/ipam-main/deploy/deploy.ps1:526:5
Line |
 526 |      New-MgOauth2PermissionGrant `
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Code: generalException Message: Location header not present in redirection response. 

INFO: Admin consent for Azure Service Management API permissions granted successfully
INFO: Deploying IPAM bicep templates
PS>TerminatingError(New-AzDeployment): "The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: 10:42:06 AM - Error: Code=InvalidTemplateDeployment; Message=The template deployment failed with multiple errors. Please see details for more information.
"
>> TerminatingError(New-AzDeployment): "The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: 10:42:06 AM - Error: Code=InvalidTemplateDeployment; Message=The template deployment failed with multiple errors. Please see details for more information.
"
ERROR: Unable to deploy Azure IPAM solution due to an exception, see ../logs/deploy_20230104104142AM.log and ../logs/error_20230104104142AM.log for detailed information!

**********************
PowerShell transcript end
End time: 20230104104206
**********************

Any assistance would be appreciated.

**Comment by DCMattyG** [comment]
Hi @mgasaway67, thank you for reaching out and I'm so sorry you're having issues getting Azure IPAM deployed.

Can you please share the output of the following PowerShell commands?

Get-InstalledModule -Name Az
Get-InstalledModule -Name Microsoft.Graph

I'd like to verify you have the minimum versions as outlined in the Prerequisites section of our documentation, which can be found here: https://azure.github.io/ipam/#/deployment/README?id=prerequisites

**Reply by mgasaway67** [reply]
I did receive some errors during one of my first deployments regarding version inconsitancies. I beleive I have correct those.  I did also get the following error at one point:

Selected Microsoft Graph profile 'v1.0' does not exist for module 'Microsoft.Graph.ManagedTenants'. No cmdlets were loaded.
I resolved this error by running the "beta" mgprofile

command ran: Select-MgProfile -Name Beta

here is the output:

PS /Users/{username}/Scripts/ipam-main/deploy> Get-InstalledModule -Name Az        

Version              Name                                Repository           Description
-------              ----                                ----------           -----------
9.2.0                Az                                  PSGallery            Microsoft Azure PowerShell - Cmdlets to manage resources in Azure. This module is com…

PS /Users/{username}/Scripts/ipam-main/deploy> Get-InstalledModule -Name Microsoft.Graph

Version              Name                                Repository           Description
-------              ----                                ----------           -----------
1.19.0               Microsoft.Graph                     PSGallery            Microsoft Graph PowerShell module

**Reply by DCMattyG** [reply]
Thanks for sharing this @mgasaway67, given the above note, are you now able to successfully deploy IPAM? I test these scripts often on Windows and Linux machines, but unfortunately I don't have a Mac machine handy to test this on a regular basis. Certainly if you're running into issues, I want to incorporate any workarounds into a troubleshooting section in the event other users run into the same.

**Reply by mgasaway67** [reply]
no, I am unable to successfully deploy.  The OP still reflects my issues.

**Reply by DCMattyG** [reply]
Can you share both of the error logs (from the logs folder) with me for review? Feel free to email them to me if that is more comfortable for you at Matthew.Garrett@microsoft.com

**Reply by mgasaway67** [reply]
sent

**Reply by DCMattyG** [reply]
Thank you so much @mgasaway67!

Can you share the output of this command in PowerShell?

bicep --version

**Comment by mgasaway67** [comment]
Bicep CLI version 0.13.1 (e3ac80d678)

**Reply by DCMattyG** [reply]
Can you add this line to the end of the script?

$PSItem.ToString() | Out-File -FilePath $errorLog -Append

Add it here:
![image](https://user-images.githubusercontent.com/12955732/210621834-0eadb84c-a72d-484d-8138-ac41bcb24248.png)

For some reason the full error isn't being written to the logs, hopefully this will provide more insights as to what is wrong. I will make the changes permanent to the deployment script once we figure out what's going on here.

**Reply by DCMattyG** [reply]
Additionally, can you provide more details on how you are invoking the deploy.ps1?

Are you just doing a simple Full deployment with?: .\deploy.ps1 -Location "&lt;location&gt;"

Or are you specifying other switches in your deployment?

**Reply by mgasaway67** [reply]
added and running the script again.

**Reply by mgasaway67** [reply]
logs sent through email.

**Reply by mgasaway67** [reply]
./deploy.ps1 -location "usgovvirginia" -UIAppName "customname" -EngineAppName "customname" -namePrefix "customprefix"

**Reply by mgasaway67** [reply]
I would add that when I deploy I am getting the appReg's created. Everytime I run I get a new AppReg of the same name. I am wondering if the API permissions consent is not working for some reason? When I look at the app reg's they are not consent'd until I manually consent them.

![image](https://user-images.githubusercontent.com/55165074/210625960-551afc97-6782-414a-8fcf-312398609449.png)

**Reply by DCMattyG** [reply]
Are you a Global Administrator?

**Reply by mgasaway67** [reply]
Yes I am. I am Azure owner and AAD GA,
